File: 0015-3sum/0015-3sum.go
import (
    "sort"
)

func threeSum(nums []int) [][]int {
     sort.Ints(nums)
    result := [][]int{}

    for i := 0; i < len(nums)-2; i++ {
        if i > 0 && nums[i] == nums[i-1] {
            continue
        }

        left, right := i + 1, len(nums) - 1
        twoSum := -nums[i]

        for left < right {
            sum := nums[left] + nums[right]

            if sum == twoSum {
                result = append(result, []int{nums[i], nums[left], nums[right]})
                left++
                right--

                for left < right && nums[left] == nums[left-1] {
                    left++
                }
                for left < right && nums[right] == nums[right+1] {
                    right--
                }

            } else if sum < twoSum {
                left++
            } else {
                right--
            }
        }
    }
    return result
}

File: 0015-3sum/0015-3sum.java
import java.util.Arrays;

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> results = new ArrayList<>();

        Arrays.sort(nums);

        for (int i = 0; i < nums.length - 2; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) continue;

            int left = i + 1;
            int right = nums.length - 1;
            
            while (left < right) {
                int sum = nums[i] + nums[left] + nums[right];

                if (sum == 0) {
                    List<Integer> elm = new ArrayList<>();
                    elm.add(nums[i]);
                    elm.add(nums[left]);
                    elm.add(nums[right]);
                    results.add(elm);

                    while (left < right && nums[left] == nums[left + 1]) left++;
                    while (left < right && nums[right] == nums[right - 1]) right--;

                    left++;
                    right--;
                } else if (sum < 0) {
                    left++;
                } else {
                    right--;
                }
            }
        }
        return results;
    }
}